libs/data-mapper/src/lib/ui/ReactFlowWrapper.tsx

```diff
@@ -152,21 +156,38 @@ export const ReactFlowWrapper = ({ canvasBlockHeight }: ReactFlowWrapperProps) =
     [nodes]
   );
 
+  // Restrict canvas panning to certain bounds
+  const translateExtent = useMemo<CanvasExtent>(() => {
+    const xOffset = schemaNodeCardWidth * 2;
+    const yOffset = schemaNodeCardHeight * 2;
+
+    const xPos = canvasBlockWidth / canvasZoom - xOffset;
+    const yPos = canvasBlockHeight / canvasZoom - yOffset;
+
+    return [
+      [-xPos, -yPos],
+      [xPos + diagramSize.width, yPos + diagramSize.height],
+    ];
+  }, [diagramSize, canvasBlockHeight, canvasBlockWidth, canvasZoom]);
+
   return (
     <ReactFlow
       ref={reactFlowRef}
-      onKeyDown={keyDownHandler}
       nodeTypes={nodeTypes}
       edgeTypes={edgeTypes}
       nodes={nodes}
       edges={edges}
+      onPaneClick={onPaneClick}
+      // Not ideal, but it's this or useViewport that re-renders 3000 (due to x/y changes)
```

Comment:
The state update that happens here only triggers the dependent useMemo if the value changes as far as I know, so it should only be recalculating the TE on zoom value changes!  We should be fine using this as is